### ISSUE TITLE
Save stack name early for clean deletions

### ISSUE THREAD

**Issue by Fuuzetsu** [tsurucapital/docker-machine-driver-cloudformation#1]
It seems driver state is not saved until we leave function like `Create`. But that means we have ample opportunity after we create the stack for the stack to fail and not end up creating an instance.

Instead, we should leave Create after creating stack and waiting for instance in `Status` or something. I'm not sure if this is viable as docker-machine may be being impatient in `Status`... I will have to try.

If there was some `saveCurrentState` function I'm missing, that'd be pretty cool too.

**Comment by Fuuzetsu** [comment]
I'm looking at the libmachine.go code and it looks like this:

```golang
	if err := h.Driver.Create(); err != nil {
		return fmt.Errorf("Error in driver during machine creation: %s", err)
	}

	if err := api.Save(h); err != nil {
		return fmt.Errorf("Error saving host to store after attempting creation: %s", err)
	}

	// TODO: Not really a fan of just checking "none" or "ci-test" here.
	if h.Driver.DriverName() == "none" || h.Driver.DriverName() == "ci-test" {
		return nil
	}

	log.Info("Waiting for machine to be running, this may take a few minutes...")
	if err := mcnutils.WaitFor(drivers.MachineInState(h.Driver, state.Running)); err != nil {
		return fmt.Errorf("Error waiting for machine to be running: %s", err)
	}
```

So it's obviously not saving state. So I think that instead what we have to do is wait until the `MachineInState` bit:
```
func MachineInState(d Driver, desiredState state.State) func() bool {
	return func() bool {
		currentState, err := d.GetState()
		if err != nil {
			log.Debugf("Error getting machine state: %s", err)
		}
		if currentState == desiredState {
			return true
		}
		return false
	}
}
```

this just calls `GetState`. So I think we

1. Never error in Create past stack creation step
2. If we have to report some failure after stack creation, we put it in the `driver` state and jump out
3. In `GetState` we check the failure and return it.

Basically we smuggle failures past the save point.

**Comment by Fuuzetsu** [comment]
For future reference, it may seem a strange way to do it: surely we could just finish `Create` after the stack is made and wait for the EC2 instance to come up as part of `GetState`, just returning `Starting` or something. This is because I was under the impression that the `mcnutils.WaitFor` wrapping `MachineInState` put some time bound on how long GetState is allowed to take But if we look at these:

```golang
func WaitForSpecific(f func() bool, maxAttempts int, waitInterval time.Duration) error {
	return WaitForSpecificOrError(func() (bool, error) {
		return f(), nil
	}, maxAttempts, waitInterval)
}

func WaitFor(f func() bool) error {
	return WaitForSpecific(f, 60, 3*time.Second)
}
```

We see that it only measures how many times it's allowed to try it. So I think the cleaner way actually works, I will try it.
